crates/ruff_linter/src/rules/flake8_quotes/rules/avoidable_escaped_quote.rs

```diff
@@ -231,13 +289,149 @@ pub(crate) fn avoidable_escaped_quote(
+                );
+                diagnostic.set_fix(Fix::safe_edit(Edit::range_replacement(
+                    fixed_contents,
+                    tok_range,
+                )));
+                diagnostics.push(diagnostic);
+            }
+            Tok::FStringStart => {
+                let text = locator.slice(tok_range);
+                // Check for escaped quote only if we're using the preferred quotation
+                // style and it isn't a triple-quoted f-string.
+                let check_for_escaped_quote = !is_triple_quote(text);
+                let quote_style = if text.contains(Quote::Single.as_char()) {
+                    Quote::Single
+                } else {
+                    Quote::Double
+                };
+                fstrings.push(FStringContext::new(
+                    check_for_escaped_quote,
+                    tok_range,
+                    quote_style,
+                ));
+            }
+            Tok::FStringMiddle {
+                value: string_contents,
+                is_raw,
+            } if !is_raw => {
+                let Some(context) = fstrings.last_mut() else {
+                    continue;
+                };
+                if !context.check_for_escaped_quote {
+                    continue;
+                }
+                if contains_escaped_quote(context.quote_style.opposite().as_char(), string_contents)
+                {
+                    context.push_fstring_middle_range(tok_range);
+                }
+            }
+            Tok::FStringEnd => {
+                let Some(context) = fstrings.pop() else {
+                    continue;
+                };
+                if context.middle_ranges_with_escapes.is_empty() {
+                    // There are no `FStringMiddle` tokens containing any escaped
+                    // quotes.
+                    continue;
+                }
+                let mut diagnostic = Diagnostic::new(
+                    UnnecessaryEscapedQuote,
+                    TextRange::new(context.start_range.start(), tok_range.end()),
+                );
+                let mut fstring_middle_edits =
+                    context.middle_ranges_with_escapes.iter().map(|&range| {
+                        Edit::range_replacement(
+                            unescape_string(
+                                locator.slice(range),
+                                context.quote_style.opposite().as_char(),
+                            ),
+                            range,
+                        )
+                    });
+                diagnostic.set_fix(Fix::safe_edits(
+                    fstring_middle_edits.next().unwrap(),
+                    fstring_middle_edits,
+                ));
+                diagnostics.push(diagnostic);
+            }
+            _ => {}
+        }
+    }
+}
+
+fn contains_escaped_quote(quote: char, haystack: &String) -> bool {
```

Comment:
This is awfully similar to `unescape_string`. Probably we could literally call this and check if the strings are identical, instead of having the same logic twice...

The performance here is probably slightly better since it avoids building a new string and bails out as soon as it found an escaped quote, but I'd be surprised if there was any noticeable difference even on a large project...